File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        maxpro = 0
        minbuy = prices[0]

        for sell in prices:
            maxp = sell - minbuy
            maxpro = max(maxp, maxpro)
            minbuy = min(sell, minbuy)

        return maxpro

# TC = O(n)
# SC = O(1)

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int l = 0;
        int r = 1;
        int profit = 0;

        while (r < prices.length) {
            if (prices[l] < prices[r]) {
                profit = Math.max(profit, prices[r] - prices[l]);
            }
            else {
                l = r;
            }
            r += 1;
        }
        return profit;
    }
}